src-extras/Database/LSMTree/Extras/MergingTreeData.hs

```diff
@@ -0,0 +1,358 @@
+  -> IO (Ref (MergingTree IO h))
+unsafeFlushMergingTree hfs hbio resolve indexType path counter = \case
+    CompletedTreeMergeData rd -> do
+      run <- unsafeFlushRun hfs hbio indexType path counter rd
+      tree <- MT.mkMergingTree . MT.CompletedTreeMerge =<< dupRef run
+      releaseRef run
+      return tree
+    OngoingTreeMergeData mrd -> do
+      mr <- unsafeFlushMergingRun hfs hbio resolve indexType path counter mrd
+      tree <- MT.mkMergingTree . MT.OngoingTreeMerge =<< dupRef mr
+      releaseRef mr
+      return tree
+    PendingLevelMergeData prds mtd -> do
+      prs <- forM prds $ \case
+        PreExistingRunData rd -> MT.PreExistingRun
+          <$> unsafeFlushRun hfs hbio indexType path counter rd
+        PreExistingMergingRunData mrd -> MT.PreExistingMergingRun
+          <$> unsafeFlushMergingRun hfs hbio resolve indexType path counter mrd
+      mt <- forM mtd $
+        unsafeFlushMergingTree hfs hbio resolve indexType path counter
+      tree <- MT.newPendingLevelMerge prs mt
+      forM_ prs $ \case
+        MT.PreExistingRun r         -> releaseRef r
+        MT.PreExistingMergingRun mr -> releaseRef mr
+      forM_ mt $ releaseRef
+      return tree
+    PendingUnionMergeData mtds -> do
+      mts <- forM mtds $
+        unsafeFlushMergingTree hfs hbio resolve indexType path counter
+      tree <- MT.newPendingUnionMerge mts
+      forM_ mts $ releaseRef
+      return tree
+
+{-------------------------------------------------------------------------------
+  MergingTreeData
+-------------------------------------------------------------------------------}
+
+-- | A data structure suitable for creating arbitrary 'MergingTree's.
+--
+-- Note: 'b ~ Void' should rule out blobs.
+data MergingTreeData k v b =
+    CompletedTreeMergeData (RunData k v b)
+  | OngoingTreeMergeData (MergingRunData MR.TreeMergeType k v b)
+  | PendingLevelMergeData
+      [PreExistingRunData k v b]
+      (Maybe (MergingTreeData k v b))  -- ^ not both empty!
+  | PendingUnionMergeData [MergingTreeData k v b]  -- ^ at least 2 children
+  deriving stock (Show, Eq)
+
+data PreExistingRunData k v b =
+    PreExistingRunData (RunData k v b)
+  | PreExistingMergingRunData (MergingRunData MR.LevelMergeType k v b)
+  deriving stock (Show, Eq)
+
+-- | See @treeInvariant@ in prototype.
+mergingTreeDataInvariant :: MergingTreeData k v b -> Either String ()
+mergingTreeDataInvariant = \case
+    CompletedTreeMergeData _rd ->
+      Right ()
+    OngoingTreeMergeData mr ->
+      mergingRunDataInvariant mr
+    PendingLevelMergeData prs t -> do
+      assertI "pending level merges have at least one input" $
+        length prs + length t > 0
+      for_ prs $ \case
+        PreExistingRunData        _r -> Right ()
+        PreExistingMergingRunData mr -> mergingRunDataInvariant mr
+      for_ t mergingTreeDataInvariant
+    PendingUnionMergeData ts -> do
+      assertI "pending union merges are non-trivial (at least two inputs)" $
+        length ts > 1
```

Comment:
Minor nit: I find `>=2` clearer, because that's also how you describe the invariant in the text